Framework/DataHandling/src/LoadSampleShape.cpp

```diff
@@ -140,6 +140,123 @@ std::unique_ptr<Geometry::MeshObject> readSTLSolid(std::ifstream &file,
   return nullptr;
 }
 
+bool getOFFline(std::ifstream &file, std::string &line) {
+  // Get line from OFF file ignoring blank lines and comments
+  // The line output is ready trimmed
+  if (!getline(file, line)) {
+    return false;
+  }
+  boost::trim(line);
+  while (line.empty() || line.substr(0, 1) == "#") {
+    if (!getline(file, line)) {
+      return false;
+    }
+    boost::trim(line);
+  }
+  return true;
+}
+
+void readOFFVertices(std::ifstream &file, uint16_t nV,
+                     std::vector<V3D> &vertices) {
+  std::string line;
+  V3D vertex;
+  for (uint16_t i = 0; i < nV; i++) {
+    if (getOFFline(file, line)) {
+      std::vector<std::string> tokens;
+      boost::split(tokens, line, boost::is_any_of(" "),
+                   boost::token_compress_on);
+      if (tokens.size() == 3) {
+        vertex.setX(boost::lexical_cast<double>(tokens[0]));
+        vertex.setY(boost::lexical_cast<double>(tokens[1]));
+        vertex.setZ(boost::lexical_cast<double>(tokens[2]));
+      } else {
+        throw std::runtime_error("Error on reading OFF vertex");
+      }
+      vertices.push_back(vertex);
+    } else {
+      throw std::runtime_error(
+          "Unexpected end of file, while reading OFF vertices");
+    }
+  }
+}
+
+void readOFFTriangles(std::ifstream &file, uint16_t nT,
+                      std::vector<uint16_t> &triangleIndices) {
+  std::string line;
+  uint16_t t1, t2, t3;
+  size_t nFaceVertices;
+  for (uint16_t i = 0; i < nT; i++) {
+    if (getOFFline(file, line)) {
+      std::vector<std::string> tokens;
+      boost::split(tokens, line, boost::is_any_of(" "),
+                   boost::token_compress_on);
+      if (tokens.size() >= 4) {
+        nFaceVertices = boost::lexical_cast<size_t>(tokens[0]);
+        if (nFaceVertices == 3) {
+          t1 = boost::lexical_cast<uint16_t>(tokens[1]);
+          t2 = boost::lexical_cast<uint16_t>(tokens[2]);
+          t3 = boost::lexical_cast<uint16_t>(tokens[3]);
+        } else {
+          throw std::runtime_error("OFF face is not a triangle.");
+        }
+        triangleIndices.push_back(t1);
+        triangleIndices.push_back(t2);
+        triangleIndices.push_back(t3);
+      } else {
+        throw std::runtime_error("Error on reading OFF triangle");
+      }
+    } else {
+      throw std::runtime_error(
+          "Unexpected end of file, while reading OFF triangles");
+    }
+  }
+}
+
+std::unique_ptr<MeshObject> readOFFMeshObject(std::ifstream &file) {
+  std::vector<uint16_t> triangleIndices;
+  std::vector<V3D> vertices;
+  uint16_t nVertices;
+  uint16_t nTriangles;
+
+  std::string line;
+  // Get number of vetrtices and faces
+  if (getOFFline(file, line)) {
+    std::vector<std::string> tokens;
+    boost::split(tokens, line, boost::is_any_of(" "), boost::token_compress_on);
+    if (tokens.size() == 3) {
+      nVertices = boost::lexical_cast<uint16_t>(tokens[0]);
```

Comment:
What happens when an instrument geometry has more than (2^16) 65536 vertices? Some ESS instruments will have more than 100K detectors which could potentially be defined as OFF files.

Reply:
The OFF format is currently being added as a way to represent instrument geometry by the Nexus Standard. I am currently working on a fork of Mantid which introduces OFF geometry using `MeshObject` so this change is imminent.

Reply:
I am happy for that to be handled in a separate PR though. Doesn't have to be dealt with here and now.

Reply:
If you believe there will be MeshObjects with more than 2^16 vertices, please raise it as an issue or make the necessary changes yourself. If you do make the necessary changes, LoadSampleShape would need to be modified after this issue merged with master.